File: 98-validate-binary-search-tree/validate-binary-search-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
// Recursive Depth-First-Search
// Time: O(n)
// Space: O(h)...h for height of tree...if the tree is balanced
var isValidBST = function(root) {
    
    const dfs = (nde, min, max) => {
        if (!nde) return true;

        if (nde.val <= min || nde.val >= max) return false;

        let left = dfs(nde.left, min, nde.val), right = dfs(nde.right, nde.val, max)

        return left && right
    }
    return dfs(root, -Infinity, Infinity)
};

File: 98-validate-binary-search-tree/validate-binary-search-tree.swift
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
class Solution {
    func isValidBST(_ root: TreeNode?) -> Bool {
        return isValidBST(root, min: Int.min, max: Int.max)
    }

    func isValidBST(_ root: TreeNode?, min: Int, max: Int) -> Bool {
        guard let root = root else { return true }
        var left: Bool = isValidBST(root.left, min: min, max: root.val)
        var right: Bool = isValidBST(root.right, min: root.val, max : max)

        return left && right && (root.val < max) && (root.val > min)
    }
}